### PR TITLE
[Feat] Token allowance management

### PR DESCRIPTION
**Pull request by fpezzola**
# Name of the feature/issue
Token allowance management

## Description
BlockWallet will automatically recognize all the token allowances the account granted to every token address and spender. Account users will also be able to edit those allowances.

This PR also fixes some issues related to the events submitted on every transaction's state transitions.

### CODE REVIEW

**Review by julianariel — APPROVED** [review]
impressive work! 🪄 🎩 

thanks @fpezzola and @antonyayoub for the huge effort 

LGTM

File: packages/background/src/utils/rpc/RPCLogsFetcher.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                    fromBlock + max * (i + 1),
+                    toBlock
+                );
+                await chunkExecutor(chunkFromBlock, chunkToBlock);
+            }
+        }
+    }
+
+    private _getBlockAsNumber(block: string | number): number {
+        return typeof block === 'string' ? parseInt(block) : block;
```
[the remark this answers is not in the document]
Comment (fpezzola) [reply packages/background/src/utils/rpc/RPCLogsFetcher.ts]: Really good catch. I have added some code to improve this. What do you think?

File: packages/background/src/utils/logsQuery.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    tokentx: id('Transfer(address,address,uint256)'),
+    token1155tx: '',
+    tokennfttx: '',
+};
+
+const TOKEN_ALLOWACE_SIGNATURES: { [type in WatchedTransactionType]: string } =
+    {
+        txlist: '',
+        tokentx: id('Approval(address,address,uint256)'),
+        token1155tx: 'ApprovalForAll(address,address,bool)',
```
[the remark this answers is not in the document]
Comment (fpezzola) [reply packages/background/src/utils/logsQuery.ts]: Good catch!

File: packages/background/src/controllers/BlankController.ts [hunk 3]
```diff
@@ -2713,6 +2799,34 @@ export default class BlankController extends EventEmitter {
...
+    }: RequestApproveAllowance): Promise<boolean> {
+        return this.tokenAllowanceController.approveAllowance(
+            BigNumber.from(allowance),
+            BigNumber.from(amount),
+            spenderAddress,
+            feeData,
+            tokenAddress,
+            customNonce
+        );
+    }
```
[the remark this answers is not in the document]
Comment (fpezzola) [reply packages/background/src/controllers/BlankController.ts:2828]: I think @antonyayoub can give us a better explanation. But here are my thoughts on this... 

First of all, let me give you a little bit of context. Before this feature, we had 3 different ways of firing "approval" transactions: 
- DApp flow (e.g Uniswap) => Tx is added to the state and then redirected to the ApproveAsset.
- Swaps flow => Tx does not exists on the store, but the user is redirected to the ApprovePage and, once submitted, the tx is created, signed and submitted.
- Bridges flow => Same as Swap flow. 

Now, with this new feature, we are adding a 4th possibility: Firing approval transactions from the Token allowance screen. This, added to the fact that we need to handle the "revoke all" functionality (multiple revoke allowance transactions queued), we decided to re-use the multiple tx functionality we have in the Dapp flow. To do so, we needed a method to add an allowance transaction to the store from the UI without "approving" it, simulating the Dapp flow.

[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/controllers/BlankController.ts:2828]: To summarise what @fpezzola explained above, the **approveAllowance** function approves the allowance by executing it on the chain directly with the needed params(feeData &  customNonce) while the **addAsNewApproveTransaction** function adds the transaction with the TransactionStatus.UNAPPROVED status so the UI can catch its details in the ApproveAssets page and let the user decide the gas fees and the confirmation. 
This was mainly needed for the "revoke all" functionality to queue the allowances back to back on the UI.

**Pull request merged**